### PR TITLE
Access token callback

### PR DESCRIPTION
**Pull request by tkyc**
Allows client code to register a callback for DataSource that returns an access token.

### CODE REVIEW

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java [hunk 1]
```diff
@@ -1187,9 +1196,9 @@ public void registerColumnEncryptionKeyStoreProvidersOnConnection(
                 }
 
                 if (null == entry.getValue()) {
-                    throw new SQLServerException(null,
-                            String.format(SQLServerException.getErrString("R_CustomKeyStoreProviderValueNull"),
-                                    providerName), null, 0, false);
+                    throw new SQLServerException(null, String
+                            .format(SQLServerException.getErrString("R_CustomKeyStoreProviderValueNull"), providerName),
+                            null, 0, false);
                 }
```
Comment (tkyc) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java:1203]: Everything above this comment is just formatting changes.

File: src/test/java/com/microsoft/sqlserver/jdbc/connection/NativeMSSQLDataSourceTest.java [hunk 2]
```diff
@@ -66,6 +84,54 @@ public void testDSNormal() throws ClassNotFoundException, IOException, SQLExcept
...
+
+        SQLServerConnectionPoolDataSource ds = new SQLServerConnectionPoolDataSource();
+        AbstractTest.updateDataSource(connectionString, ds);
+        ds.setUser("");
+        ds.setPassword("");
+        ds.setAccessTokenCallback(callback);
+
+        // Callback should provide valid token on connection open for all new connections
+        try (Connection conn1 = (SQLServerConnection) ds.getConnection()) {}
+        try (Connection conn2 = (SQLServerConnection) ds.getConnection()) {}
```
Comment (David-Engel) [inline src/test/java/com/microsoft/sqlserver/jdbc/connection/NativeMSSQLDataSourceTest.java:132]: There are a few more validations you can do here.

1. When the access token hasn't expired, validate that the underlying connection is the same between conn1 and conn2.
2. When the access token has expired (manually set the date to something that will trigger checkIfNeedNewAccessToken in Util.java), validate that the underlying connection is different between conn1 and conn2.
3. Connect, run a query, kill the connection and run another query. ICR should reconnect and get an access token using the callback.

File: src/test/java/com/microsoft/sqlserver/testframework/AbstractTest.java [hunk 3]
```diff
@@ -138,6 +143,9 @@ public static void setup() throws Exception {
         applicationKey = getConfiguredProperty("applicationKey");
         tenantID = getConfiguredProperty("tenantID");
 
+        accessTokenClientId = getConfiguredProperty("accessTokenClientId");
+        accessTokenSecret = getConfiguredProperty("accessTokenSecret");
```
Comment (David-Engel) [inline src/test/java/com/microsoft/sqlserver/testframework/AbstractTest.java:147]: Can we re-use the existing principal ID and secret we already have?

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java [hunk 4]
```diff
@@ -5542,7 +5588,7 @@ private SqlFedAuthToken getFedAuthToken(SqlFedAuthInfo fedAuthInfo) throws SQLSe
         int sleepInterval = 100;
 
         if (!msalContextExists()
-                && !authenticationString.equalsIgnoreCase(SqlAuthentication.ActiveDirectoryInteractive.toString())) {
+                && !authenticationString.equalsIgnoreCase(SqlAuthentication.ActiveDirectoryIntegrated.toString())) {
```
[the remark this answers is not in the document]
Comment (lilgreenbird) [reply src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java:5591]: there should also be a check for ActiveDirectoryInteractive that needs MSAL as well. I think previously there's this and also a check for ActiveDirectoryIntegrated in another place

File: src/main/java/com/microsoft/sqlserver/jdbc/SqlAuthenticationToken.java [hunk 5]
```diff
@@ -10,30 +10,39 @@
...
      */
     private static final long serialVersionUID = -1343105491285383937L;
 
-    final Date expiresOn;
-    final String accessToken;
+    private final Date expiresOn;
+    private final String accessToken;
 
-    SqlFedAuthToken(String accessToken, long expiresIn) {
+    public SqlAuthenticationToken(String accessToken, long expiresIn) {
```
Comment (lilgreenbird) [inline src/main/java/com/microsoft/sqlserver/jdbc/SqlAuthenticationToken.java:25]: javadocs, same in line 33, pls check all build warnings

File: src/samples/azureactivedirectoryauthentication/src/main/java/AzureActiveDirectoryAccessTokenCallback.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
+/*
+ * Microsoft JDBC Driver for SQL Server Copyright(c) Microsoft Corporation All rights reserved. This program is made
+ * available under the terms of the MIT License. See the LICENSE file in the project root for more information.
+ */
+package azureactivedirectoryauthentication.src.main.java;
```
Comment (lilgreenbird) [inline src/samples/azureactivedirectoryauthentication/src/main/java/AzureActiveDirectoryAccessTokenCallback.java:5]: why is this in another package?

File: src/test/java/com/microsoft/sqlserver/testframework/AbstractTest.java (hunk 3 above)
Comment (tkyc) [reply src/test/java/com/microsoft/sqlserver/testframework/AbstractTest.java:147]: Yes, my mistake. After checking, there already is a pair secret/clientId available.

**Comment by Jeffery-Wasty** [comment]
Multiple comments have leading hyphens, I don't think we do that anywhere. Please remove them for consistency.

File: src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerConnection.java [hunk 7]
```diff
@@ -411,13 +411,18 @@ CallableStatement prepareCall(String sql, int nType, int nConcur, int nHold,
...
      */
     @Deprecated
     int getMsiTokenCacheTtl();
 
     /**
      * Deprecated. Time-to-live is no longer supported for the cached Managed Identity tokens.
      * This method is a no-op for backwards compatibility only.
+     *
+     * @param timeToLive
+     *        - Time-to-live is no longer supported.
```
Comment (Jeffery-Wasty) [inline src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerConnection.java:425]: This is an example of what I mean.

File: src/samples/azureactivedirectoryauthentication/src/main/java/AzureActiveDirectoryAccessTokenCallback.java [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+        ds.setServerName("<your-server-name>");
+        ds.setDatabaseName("<your-database-name>");
+        ds.setAccessTokenCallback(callback);
+
+        try (Connection conn = (SQLServerConnection) ds.getConnection()) {
+            System.out.println("Connected...");
+        }
+
+    }
+}
```
Comment (Jeffery-Wasty) [inline src/samples/azureactivedirectoryauthentication/src/main/java/AzureActiveDirectoryAccessTokenCallback.java:66]: Does this need to be added to the pom in `\src\samples\azureactivedirectoryauthentication`?

Comment (tkyc) [reply src/samples/azureactivedirectoryauthentication/src/main/java/AzureActiveDirectoryAccessTokenCallback.java:66]: Updated pom

File: src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerDataSource.java [hunk 9]
```diff
@@ -1216,14 +1216,34 @@ public interface ISQLServerDataSource extends javax.sql.CommonDataSource {
     /**
      * Deprecated. Time-to-live is no longer supported for the cached Managed Identity tokens.
      * This method is a no-op for backwards compatibility only.
+     *
+     * @param timeToLive
+     *        - Time-to-live is no longer supported.
```
Comment (Jeffery-Wasty) [inline src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerDataSource.java:1221]: missed one

**Review by Jeffery-Wasty — DISMISSED** [review]

**Review by lilgreenbird — APPROVED** [review]

**Review by Jeffery-Wasty — APPROVED** [review]

**Review by Jeffery-Wasty — APPROVED** [review]

**Pull request merged**